[TensorExpr] Add lowering for aten::max (reduction).

ZolotukhinM opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #66847
* #66846
* #66527
* #66522
* #66515
* #66514
* #66523
* __->__ #66519
* #66518
* #66517
* #66516



Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

Commit:
[TensorExpr] Add lowering for aten::max (reduction).

Is it a hack?
Yes and no. Yes, because I had to disable lowering for pointwise max.
No, because eventually we would need to add max reduction lowering
anyway. A proper implementation requires #65843 - that would allow us to
distinguish pointwise case from the reduction case and have separate
lowerings for them. In the long run we should land #65843 and then a
cleaned up version of this PR.

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590853).

Commit:
Update on "[TensorExpr] Add lowering for aten::max (reduction)."


Is it a hack?
Yes and no. Yes, because I had to disable lowering for pointwise max.
No, because eventually we would need to add max reduction lowering
anyway. A proper implementation requires #65843 - that would allow us to
distinguish pointwise case from the reduction case and have separate
lowerings for them. In the long run we should land #65843 and then a
cleaned up version of this PR.

Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add lowering for aten::max (reduction)."


Is it a hack?
Yes and no. Yes, because I had to disable lowering for pointwise max.
No, because eventually we would need to add max reduction lowering
anyway. A proper implementation requires #65843 - that would allow us to
distinguish pointwise case from the reduction case and have separate
lowerings for them. In the long run we should land #65843 and then a
cleaned up version of this PR.

Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add lowering for aten::max (reduction)."


Is it a hack?
Yes and no. Yes, because I had to disable lowering for pointwise max.
No, because eventually we would need to add max reduction lowering
anyway. A proper implementation requires #65843 - that would allow us to
distinguish pointwise case from the reduction case and have separate
lowerings for them. In the long run we should land #65843 and then a
cleaned up version of this PR.

Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

[ghstack-poisoned]

Commit:
Update on "[TensorExpr] Add lowering for aten::max (reduction)."


Is it a hack?
Yes and no. Yes, because I had to disable lowering for pointwise max.
No, because eventually we would need to add max reduction lowering
anyway. A proper implementation requires #65843 - that would allow us to
distinguish pointwise case from the reduction case and have separate
lowerings for them. In the long run we should land #65843 and then a
cleaned up version of this PR.

Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590853).

eellison reviewed (commented):
How close are we to codegen'ing this ? I know @navahgar  was at one point looking at reductions

ZolotukhinM commented:
> How close are we to codegen'ing this ? I know @navahgar was at one point looking at reductions

I think we can codegen it now, it's just safer to go with an external call in the beginning since it would work for sure (whereas if I try to go with a native lowering I might screw up perf and/or correctness).

Commit:
Update on "[TensorExpr] Add lowering for aten::max (reduction)."




Differential Revision: [D31590853](https://our.internmc.facebook.com/intern/diff/D31590853)

[ghstack-poisoned]

ZolotukhinM commented:
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D31590853).

navahgar commented on torch/csrc/jit/tensorexpr/external_functions.cpp:
```diff
@@ -384,6 +384,28 @@ void nnc_aten_mean(
+  at::Tensor& r = tensors[0];
+  const at::Tensor& x = tensors[1];
+  int64_t max_dim = extra_args[0];
+  bool keep_dim = extra_args[1];
```
Do we need a cast to `bool` here? Just to be explicit.

navahgar reviewed (approved):
LGTM
